fix tensor print behavior for XPU

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130523

# Motivation
Some XPU device don't support `double` data type. So we have to use `tensor.to(torch.float)` if it is a XPU tensor.


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

guangyey commented:
@pytorchbot rebase -b main

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
fix tensor print behavior for XPU

[ghstack-poisoned]

albanD commented on torch/_tensor_str.py:
```diff
@@ -115,7 +115,7 @@ def printoptions(**kwargs):
 
 def tensor_totype(t):
-    dtype = torch.float if t.is_mps else torch.double
+    dtype = torch.float if (t.is_mps or t.is_xpu) else torch.double
```
You don't have double precision dtypes?

guangyey replied:
Let me clarify that, in general, enterprise-grade GPU have double precision types, while consumer-grade GPU do not.

dvrogozh replied:
If you will change precision here, then some calculations might start overflow and final printout might confuse users. Also, as you just commented, some of XPU devices has fp64 support, some - do not. Thus, checking for `is_xpu` is not a fine grained check.

I would go with double as much as possible. For XPU we can have a finer grain check using `torch.xpu.get_device_properties().has_fp64`.

gujinghui replied:
It's better if we can add the check of has_fp64 here. @guangyey  pls. help confirm.

BTW, it should be safe if not convert to double, as most of low precision float data types have been converted to float already before print?
https://github.com/pytorch/pytorch/blob/168e41009bbd664c805a6078ed20d44ffe101335/torch/_tensor_str.py#L332

albanD replied:
> BTW, it should be safe if not convert to double, as most of low precision float data types have been converted to float already before print?

because we do reduction to compute extremum value, we have been on the safe side of using double to reduce the change of floating point artifacts.

guangyey commented:
@pytorchbot rebase -b main

guangyey committed:
Update

[ghstack-poisoned]

guangyey commented:
@albanD Could I address your comment?

guangyey committed:
Update on "fix tensor print behavior for XPU"


# Motivation
Some XPU device don't support `double` data type. So we have to use `tensor.to(torch.float)` if it is a XPU tensor.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good to me. Thanks for the update

gujinghui commented:
@pytorchbot merge
